File: src/hash/getMerkleProof.ts
// Copyright (C) 2022 Deliberative Technologies P.C.
// SPDX-License-Identifier: Apache-2.0
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// 	http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import sha512 from "./sha512";

import dcryptoMemory from "./memory";

import dcryptoMethodsModule from "../c/build/dcryptoMethodsModule";

import { crypto_hash_sha512_BYTES } from "../utils/interfaces";

/**
 * @function
 * Returns the Merkle proof of an element of a tree.
 * Can be used as a receipt of a transaction etc.
 *
 * @param tree: The tree.
 * @param element: The element.
 * @param serializer: Converts leaves into Uint8Array.
 *
 * @returns Promise<Uint8Array>
 */
const getMerkleProof = async <T extends Uint8Array | unknown>(
  tree: T[],
  element: T,
  serializer?: (i: T) => Uint8Array,
): Promise<Uint8Array> => {
  const treeLen = tree.length;
  if (treeLen === 0) {
    throw new Error("Cannot calculate Merkle proof of element of empty tree.");
  } else if (treeLen === 1) {
    throw new Error(
      "No point in calculating proof of a tree with single leaf.",
    );
  }

  const wasmMemory = dcryptoMemory.getMerkleProofMemory(treeLen);
  const module = await dcryptoMethodsModule({
    wasmMemory,
  });

  const ptr1 = module._malloc(treeLen * crypto_hash_sha512_BYTES);
  const leavesHashed = new Uint8Array(
    module.HEAP8.buffer,
    ptr1,
    treeLen * crypto_hash_sha512_BYTES,
  );

  let i = 0;
  let leafIsUint8Array = false;
  let hash: Uint8Array;
  let serialized: Uint8Array;
  for (const leaf of tree) {
    leafIsUint8Array =
      ArrayBuffer.isView(leaf) && leaf.constructor.name === "Uint8Array";
    if (!serializer && leafIsUint8Array) {
      hash = await sha512(leaf as Uint8Array, module);
      leavesHashed.set([...hash], i * crypto_hash_sha512_BYTES);
    } else if (serializer && !leafIsUint8Array) {
      serialized = serializer(leaf);
      hash = await sha512(serialized, module);
      leavesHashed.set([...hash], i * crypto_hash_sha512_BYTES);
    }
    // Cannot happen due to typeguards
    // else if (serializer && leafIsUint8Array) {
    //   throw new Error(
    //     "Did not need to provide a serializer since leaf is Uint8Array",
    //   );
    // }
    else {
      throw new Error("Tree leaf not Uint8Array, needs serializer.");
    }
    i++;
  }

  const ptr2 = module._malloc(crypto_hash_sha512_BYTES);
  const elementHash = new Uint8Array(
    module.HEAP8.buffer,
    ptr2,
    crypto_hash_sha512_BYTES,
  );
  leafIsUint8Array =
    ArrayBuffer.isView(element) && element.constructor.name === "Uint8Array";
  if (!serializer && leafIsUint8Array) {
    hash = await sha512(element as Uint8Array);
    elementHash.set([...hash]);
  } else if (serializer && !leafIsUint8Array) {
    serialized = serializer(element);
    hash = await sha512(serialized);
    elementHash.set([...hash]);
  }
  // Cannot happen due to typeguards
  // else if (serializer && leafIsUint8Array) {
  //   throw new Error(
  //     "Did not need to provide a serializer since element is Uint8Array",
  //   );
  // }
  // Cannot happen due to typeguards from tree
  // else {
  //   throw new Error("Element not Uint8Array, needs serializer.");
  // }

  const ptr3 = module._malloc(treeLen * (crypto_hash_sha512_BYTES + 1));
  const proof = new Uint8Array(
    module.HEAP8.buffer,
    ptr3,
    treeLen * (crypto_hash_sha512_BYTES + 1),
  );

  const result = module._get_merkle_proof(
    treeLen,
    leavesHashed.byteOffset,
    elementHash.byteOffset,
    proof.byteOffset,
  );

  module._free(ptr1);
  module._free(ptr2);

  switch (result) {
    case -1:
      module._free(ptr3);
      throw new Error("Element not in tree.");

    default: {
      const proofArray = Uint8Array.from([...proof.slice(0, result)]);
      module._free(ptr3);

      return proofArray;
    }
  }
};

export default getMerkleProof;

File: src/c/hash/get_merkle_proof.c
// Copyright (C) 2022 Deliberative Technologies P.C.
// SPDX-License-Identifier: Apache-2.0
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// 	http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include <math.h>
#include <stdbool.h>
#include <stdint.h>
#include <string.h>

#include "../utils/utils.h"

#include "../../../libsodium/src/libsodium/include/sodium/crypto_hash_sha512.h"

// The result is the proof length
__attribute__((used)) int
get_merkle_proof(
    const int LEAVES_LEN,
    const uint8_t leaves_hashed[LEAVES_LEN * crypto_hash_sha512_BYTES],
    const uint8_t element_hash[crypto_hash_sha512_BYTES],
    uint8_t proof[LEAVES_LEN * (crypto_hash_sha512_BYTES + 1)])
{
  size_t i, j, k;

  int32_t index[1];
  items_indexes_in_array(LEAVES_LEN, 1, leaves_hashed, element_hash, index);
  if (index[0] == -1) return -1;

  int element_of_interest = index[0];

  uint8_t *hashes = malloc(LEAVES_LEN * crypto_hash_sha512_BYTES);
  memcpy(hashes, leaves_hashed, LEAVES_LEN * crypto_hash_sha512_BYTES);

  uint8_t *concat_hashes = malloc(2 * crypto_hash_sha512_BYTES);
  uint8_t *hash = malloc(crypto_hash_sha512_BYTES);

  int l;
  int leaves = LEAVES_LEN;
  int res;
  bool oddLeaves;
  // Counts the index of proof artifacts.
  k = 0;
  do
  {
    // Counts the index of the elements of the new array of hashes.
    j = 0;
    oddLeaves = leaves % 2 != 0;
    for (i = 0; i < leaves; i += 2)
    {
      if (oddLeaves && i + 1 == leaves)
      {
        memcpy(concat_hashes, &hashes[i * crypto_hash_sha512_BYTES],
               crypto_hash_sha512_BYTES);
        memcpy(&concat_hashes[crypto_hash_sha512_BYTES],
               &hashes[i * crypto_hash_sha512_BYTES], crypto_hash_sha512_BYTES);

        if (i == element_of_interest)
        {
          memcpy(&proof[k * (crypto_hash_sha512_BYTES + 1)],
                 &hashes[i * crypto_hash_sha512_BYTES],
                 crypto_hash_sha512_BYTES);
          // We do not care if left(0) or right(1) since hash of itself
          proof[k++ * (crypto_hash_sha512_BYTES + 1) + crypto_hash_sha512_BYTES]
              = 0;
          element_of_interest = j;
        }
      }
      else
      {
        memcpy(concat_hashes, &hashes[i * crypto_hash_sha512_BYTES],
               crypto_hash_sha512_BYTES);
        memcpy(&concat_hashes[crypto_hash_sha512_BYTES],
               &hashes[(i + 1) * crypto_hash_sha512_BYTES],
               crypto_hash_sha512_BYTES);

        if (i == element_of_interest)
        {
          memcpy(&proof[k * (crypto_hash_sha512_BYTES + 1)],
                 &hashes[(i + 1) * crypto_hash_sha512_BYTES],
                 crypto_hash_sha512_BYTES);
          // Proof artifact needs to go to the right when concatenated with
          // element.
          proof[k++ * (crypto_hash_sha512_BYTES + 1) + crypto_hash_sha512_BYTES]
              = 1;
          element_of_interest = j;
        }
        else if (i + 1 == element_of_interest)
        {
          memcpy(&proof[k * (crypto_hash_sha512_BYTES + 1)],
                 &hashes[i * crypto_hash_sha512_BYTES],
                 crypto_hash_sha512_BYTES);
          // Proof artifact needs to go to the left when concatenated with
          // element.
          proof[k++ * (crypto_hash_sha512_BYTES + 1) + crypto_hash_sha512_BYTES]
              = 0;
          element_of_interest = j;
        }
      }

      crypto_hash_sha512(hash, concat_hashes, 2 * crypto_hash_sha512_BYTES);
      memcpy(&hashes[j++ * crypto_hash_sha512_BYTES], hash,
             crypto_hash_sha512_BYTES);
    }

    int l = ceil(leaves / 2);
    memset(&hashes[l * crypto_hash_sha512_BYTES], 0,
           (LEAVES_LEN - l) * crypto_hash_sha512_BYTES);
    leaves = l;
  } while (leaves > 1);

  free(hashes);
  free(concat_hashes);
  free(hash);

  return k * (crypto_hash_sha512_BYTES + 1);
}
